alembic/versions/614911daf883_add_seasonal_to_time_resolution_enum.py

```diff
@@ -0,0 +1,145 @@
+revision = '614911daf883'
+down_revision = None
+branch_labels = None
+depends_on = None
+
+
+connection = op.get_bind()
+dialect = connection.dialect.name
+
+
+# This is a *highly* customized migration script, barely related to the
+# autogenerated script at all. Two reasons:
+#
+# 1) There are a lot of differences between the (test, and presumably prod)
+#    database as it is now and the SQLAlchemy declaration. We ditch almost all
+#    those differences because we are frightened little creatures who don't
+#    want to wake monsters.
+# 2) Enum type. See below.
+#
+# Migration of an Enum type must be done differently depending on the database
+# type. ~sigh~
+#
+# For SQLite, we must use "batch mode" to
+# do most table alterations, including the one that changes the Enum type
+# for time_resolution. See http://alembic.zzzcomputing.com/en/latest/batch.html
+#
+# That documentation says
+#   On other backends, we’d see the usual ALTER statements done as though
+#   there were no batch directive - the batch context by default only does
+#   the “move and copy” process if SQLite is in use, and if there are migration
+#   directives other than Operations.add_column() present, which is the one
+#   kind of column-level ALTER statement that SQLite supports.
+#
+# But unfortunately Enum type is a special case, and batch mode only works for
+# SQLite.
+#
+# For PostgreSQL, we must use a different technique, as below.
+#
+# TESTING:
+#
+# These migrations have been tested on a SQLite test database and on
+# a full copy of the production database. Both upgrade and downgrade work
+# on both databases.
+
+old_options = (
+    '1-minute', '2-minute', '5-minute', '15-minute', '30-minute',
+    '1-hourly', '3-hourly', '6-hourly', '12-hourly', 'daily',
+    'monthly', 'yearly', 'other', 'irregular',
+)
+new_options = (
+    '1-minute', '2-minute', '5-minute', '15-minute', '30-minute',
+    '1-hourly', '3-hourly', '6-hourly', '12-hourly', 'daily',
+    'monthly', 'seasonal', 'yearly', 'other', 'irregular',
+)
+
+
+def alter_column(curr_options, dest_options):
+    """
+    Alter columns, which in SQLite alters the check constraint. Yay.
+    See http://alembic.zzzcomputing.com/en/latest/batch.html
+
+    :param curr_options: tuple of options (members) in the current enum type
+    :param dest_options: tuple of options (members) in the destination enum type
+    :return: None
+    """
+    old_type = sa.Enum(*curr_options, name='timescale')
+    new_type = sa.Enum(*dest_options, name='timescale')
+    with op.batch_alter_table('time_sets') as batch_op:
+        batch_op.alter_column(
+            'time_resolution', type_=new_type, existing_type=old_type)
+
+
+def swap_and_drop(curr_options, dest_options):
+    """
+    Create a temporary enum type with the destination options in it,
+    swap it for the current enum type with the current options in it,
+    then rename the temporary and drop the temporary and old enum types.
+    This works for PostgreSQL.
```

Comment:
Another good question. PG 9.1 introduced the `ALTER TYPE` for _adding_ values to enum types. There's no statement for removing them (presumably because that could invalidate column values in the database). So the upgrade (only) could be simplified, but the downgrade would still have to be as it is (with the data migration to prevent invalidation). I chose to use the same code for both operations.